sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,563 @@
+`range_kernel` property:
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+struct my_iota;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1, my_iota>)
+void iota(float start, float *ptr) {
+   // ...
+}
+```
+
+=== Interaction with the kernel bundle APIs
+
+Raw kernels work with the kernel bundle APIs defined in section 4.11 "Kernel
+bundles" of the core SYCL specification in a natural way.
+Any kernel bundle API that accepts a kernel's type-name may be used with the
+type-name of a raw kernel.
+In fact, an application must use the kernel bundle APIs in order to invoke a
+raw kernel because the application must first obtain a `kernel` object.
+Typically, the application first obtains a kernel bundle from the raw kernel's
+name and then obtains its `kernel` object as shown in the following example:
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+struct my_iota;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1, my_iota>)
+void iota(float start, float *ptr) {
+   // ...
+}
+
+int main() {
+  sycl::queue q;
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the raw kernel "my_iota".
+  auto exe_bndl =
+    sycl::get_kernel_bundle<my_iota, sycl::bundle_state::executable>(ctxt);
+
+  // Get a kernel object for the "my_iota" function from that bundle.
+  sycl::kernel myiota = exe_bndl.get_kernel<my_iota>();
+}
+```
+
+=== Enqueuing a raw kernel and setting parameter values
+
+Once the application obtains a `kernel` object for a raw kernel, it can enqueue
+the kernel to a device using any of the SYCL functions that allow a kernel to
+be enqueued via a `kernel` object.
+The application must enqueue the raw kernel according to its type.
+For example, a raw kernel defined via `range_kernel` can be enqueued by calling
+the `handler::parallel_for` overload taking a `range`.
+A raw kernel defined via `nd_range_kernel` can be enqueued by calling the
+`handler::parallel_for` overload taking an `nd_range`.
+A raw kernel defined via `single_task_kernel` can be enqueued by calling
+`handler::single_task`.
+
+Attempting to enqueue a raw kernel using a mechanism that does not match its
+type results in undefined behavior.
+Attempting to enqueue a raw kernel with a `range` or `nd_range` whose
+dimensionality does not match the raw kernel definition results in undefined
+behavior.
+
+The application is also responsible for setting the values of any kernel
+arguments when the kernel is enqueued.
+For example, when enqueuing a kernel with `handler::parallel_for` or
+`handler::single_task`, the kernel argument values must be set via
+`handler::set_arg` or `handler::set_args`.
+Failing to set the value of a kernel argument results in undefined behavior.
+
+=== Obtaining the iteration id for a kernel
+
+In a standard SYCL kernel, the iteration ID is passed as a parameter to the
+kernel's callable object.
+However, this is not the case for a raw kernel because the function parameters
```

Comment:
Why? I would prefer if the item/nd_item is the first argument followed by the other function parameters.
This would also remove the need for the range/nd_range/single_task_kernel property. Because this could be inferred from the first argument.

Reply:
Wouldn't we still need to decorate the function somehow, though?  I don't think we want to turn every function which accepts a `sycl::nd_item` into a kernel -- it might just be a device function that's called from a kernel.

Reply:
I agree we can't use the first argument to determine whether it is a kernel. Only the type of kernel (nd/range/single_task). In the same TU we can turn every function which gets passed to get_kernel into a kernel. Cross TU we do need a decoration. But then the user only needs to provide the decoration  cross TU (like SYCL_EXTERNAL) and the FE doesn't need to understand multiple decorations.